ipa-core/src/protocol/ipa_prf/quicksort.rs

```diff
@@ -71,43 +68,28 @@ where
         // check whether sort is needed
         if b_l + 1 < b_r {
             // set up iterator
-            let mut iterator = list[b_l..b_r].iter();
+            let mut iterator = list[b_l..b_r].iter().map(get_key);
             // first element is pivot, apply key extraction function f
-            let pivot = f(iterator.next().unwrap());
+            let pivot = iterator.next().unwrap();
             // create pointer to context for moving into closure
             let pctx = &(ctx.set_total_records(b_r - (b_l + 1)));
-            let pf = &f;
             // precompute comparison against pivot and reveal result in parallel
-            comp.clear();
-            seq_join(
+            let comp: BitVec<usize, Lsb0> = seq_join(
                 ctx.active_work(),
-                stream_iter(iterator.enumerate().map(|(n, x)| {
-                    async move {
-                        // compare current element against pivot
-                        let sh_comp = compare_gt(
-                            pctx.narrow(&Step::Compare),
-                            RecordId::from(n),
-                            // apply key extraction function f to element x
-                            pf(x),
-                            pivot,
-                        )
-                        .await?;
-
-                        // reveal outcome of comparison
-                        Ok::<bool, Error>(
-                            // desc = true will flip the order of the sort
-                            Boolean::from(false ^ desc)
-                                == sh_comp
-                                    .reveal(pctx.narrow(&Step::Reveal), RecordId::from(n))
-                                    .await?,
-                        )
-                    }
+                stream_iter(iterator.enumerate().map(|(n, k)| async move {
+                    // Compare the current element against pivot and reveal the result.
+                    let comparison =
+                        compare_gt(pctx.narrow(&Step::Compare), RecordId::from(n), k, pivot)
+                            .await?
+                            .reveal(pctx.narrow(&Step::Reveal), RecordId::from(n))
+                            .await?;
+
+                    // reveal outcome of comparison
+                    // desc = true will flip the order of the sort
+                    Ok::<_, Error>(Boolean::from(false ^ desc) == comparison)
                 })),
             )
-            .try_for_each(|x| {
-                comp.push(x);
-                future::ready(Ok(()))
-            })
+            .try_collect()
```

Comment:
@martinthomson I agree this is simpler. I have been working with @akoshelev on this. The reason why we went with the more complicated version is that it saves allocations. To my understanding Try_collect causes a reallocation of the bitvec (that's why I implemented it by hand and allocated the bitvec outside of the iterator).

Reply:
Given that your `BitVec` is truly tiny, I don't see this as a cost that needs to be avoided.  And log_2(N) allocations here is hardly the largest cost we have to worry about.  And in time, perhaps, the compiler can learn that multiple calls to collect() in a loop can use the same memory allocation.

Reply:
I agree in this case. Ben Savage actually merged an improvement to the quicksort implementation (using a breadth first quicksort parallelized over the recursion tree level and all user pseudonyms. I hope this resolves the main concerns with quicksort since it does not require that many different contexts with tiny record numbers anymore.) In that case, the bitvec will actually have up to query size many bits, so it might be in the range of a megabyte (for query size ~ 10 million)